### PR TITLE
Items

### PR DESCRIPTION
**Pull request by Trymunx**
Show items in context menu when right clicking, and show items around the player in the surroundings pane on the right. Currently has creatures dropping all items 20% of the time for debugging, which will need to be turned off before merging.

### CODE REVIEW

File: src/vuex/store.js [hunk 1]
```diff
@@ -37,46 +39,99 @@ const store = new Vuex.Store({
...
+        total: 0,
+        stacked: {},
+      };
       for (let y = -radius; y <= radius; y++) {
         for (let x = -radius; x <= radius; x++) {
           let tile = state.world.getTile(state.player.pos.x + x, state.player.pos.y + y);
           if (tile.items.length) {
-            surr.items.push(...tile.items);
+            items.total += tile.items.length;
+            tile.items.forEach(item => {
```
Comment (Trymunx) [inline src/vuex/store.js:75]: This whole forEach loop feels like unbelievably hacky code. It works, but I don't want it too. I'd appreciate if you could give me some pointers as to how I could better structure this data for using in the surroundings pane @Alpvax

File: src/vuex/store.js [hunk 2]
```diff
@@ -37,46 +39,99 @@ const store = new Vuex.Store({
...
+                      count: 1,
+                      totalValue: item.val,
+                      dir: getDir(x, y),
+                      loc: {[[state.player.pos.x + x, state.player.pos.y + y]]: {}},
+                    },
+                  },
+                };
+              }
+            });
+            // let tileItems = tile.items.reduce((obj, el) => {
```
Comment (Trymunx) [inline src/vuex/store.js]: This comment needs deleting

File: src/game/items/gameItems.js [hunk 3]
```diff
@@ -0,0 +1,130 @@
+import Food from "./Food";
```
Comment (Trymunx) [inline src/game/items/gameItems.js]: Need to implement the single item file to store every file, and just create them with the data they need to create a basic ECS. I'm not quite sure what the registration for each of those objects that needs to happen in this file would look like.

File: src/components/Overview.vue [hunk 4]
```diff
@@ -35,20 +35,64 @@ export default {
...
+              entity: `Examine ${creature.name}:`,
               message: `The ${creature.name} is level ${creature.level}.`
-                + ` It will drop ${creature.gold !== 0 ? creature.gold + " gold" : "nothing"}.`,
+                + ` It will drop ${outputs.length !== 0 ? output : "nothing"}.`,
             });
           },
         }
       });
       }
+      if (itemsOnTile && itemsOnTile.length > 0) {
```
Comment (Trymunx) [inline src/components/Overview.vue]: This if block doesn't seem nice.

File: src/vuex/store.js (hunk 1 above)
Comment (Alpvax) [reply src/vuex/store.js:75]: This is really hard to follow, I'm not even sure what structure you have here, or what you are after.

File: src/components/Overview.vue [hunk 5]
```diff
@@ -35,20 +35,64 @@ export default {
...
+              text: `1 ${item.name} (${item.val})`,
+              // Have to use old school function so that this points to this object
+              action: function() {console.log(this)},
+            };
+            arr.push(newItemEntry);
+          }
+          return arr;
+        }, []);
+      }
+      return (creaturesOnTile || []).concat(itemsOnTile || []);
```
[the remark this answers is not in the document]
Comment (Trymunx) [reply src/components/Overview.vue]: What if creaturesOnTile is undefined here though? Edit: never mind I just needed to do the same for the creatures getter

File: src/components/Surroundings.vue [hunk 6]
```diff
@@ -41,6 +51,11 @@ import { levelColour } from "../game/utils/colours";
 import display from "../game/overview/Display";
 
 export default {
+  data() {
+    return {
+      expandedItem: "",
```
[the remark this answers is not in the document]
Comment (Trymunx) [reply src/components/Surroundings.vue:72]: I use vuex for things that need to be accessed in multiple places by multiple files. This is only necessary for this specific component to display, so seems like a good place to just use the component's data property.

File: src/game/items/gameItems.js (hunk 3 above)
Comment (Trymunx) [reply src/game/items/gameItems.js]: Yeah that's a good shout

File: src/vuex/store.js (hunk 1 above)
Comment (Trymunx) [reply src/vuex/store.js:75]: I agree. I just kept adding bits until it worked, hence the spaghetti.
What I want at the end is an object
```{
stacked: {
  bone: {
    count: 3,
    expanded: {
      -1,1,bone: {
        count: 3,
        dir: "south west",
        loc: {
          5,8: {}
        },
        name: "bone",
        plural: "bones",
        totalValue: 75,
      },
    },
    locations: {5,8: {…}},
    name: "bone",
    plural: "bones",
  dragon blood: {
    count: 1,
    expanded: {2,-2,dragon blood: {…}},
    locations: {8,5: {…}},
    name: "dragon blood",
    plural: "dragon blood",
  },
  dragon claws: {name: "dragon claws", plural: "dragon claws", count: 7, locations: {…}, expanded: {…}},
  dragon hide: {name: "dragon hide", plural: "dragon hides", count: 2, locations: {…}, expanded: {…}},
  dragon meat: {name: "dragon meat", plural: "dragon meat", count: 2, locations: {…}, expanded: {…}},
  dragon scales: {name: "dragon scales", plural: "dragon scales", count: 6, locations: {…}, expanded: {…}},
  dragon skull: {name: "dragon skull", plural: "dragon skulls", count: 1, locations: {…}, expanded: {…}},
  dragon teeth: {name: "dragon teeth", plural: "dragon teeth", count: 6, locations: {…}, expanded: {…}},
  skull: {name: "skull", plural: "skulls", count: 1, locations: {…}, expanded: {…}},
},
total: 29,
}
```
That is, I want the total number of items in the surroundings, given by `total`. I then want a list of each of the items in the surroundings, which I store in an object called `stacked` with the item names as the keys so I can stack them (I check if that item name already exists in my `stacked` object).
Then, for each `stacked` object, I store the item's `name`, `plural` name, the number in the surroundings (`count`), the `locations` as an array so I can pass it into the highlight function in the store, and an `expanded` object.
The `expanded` object expands all of the objects with that name in the surroundings into a list of each item separated per tile, so the keys I use for that are the relative location from the player and the name, e.g. `-1,1,bone`, which JS stringifies from a 3-length array for me. As I type this, I realise that because of how I'm nesting these objects, I don't need the item name in the key, but it doesn't really matter.

Comment (Alpvax) [reply src/vuex/store.js:75]: I think using a custom type would be better here, the `expanded` seems to be a lot of duplication.

Comment (Trymunx) [reply src/vuex/store.js:75]: What do you mean by that? Make it a type instead of an object?

Comment (Trymunx) [reply src/vuex/store.js:75]: What doesn't work. Your suggestion or my code?

Comment (Alpvax) [reply src/vuex/store.js:75]: My suggestion. Vue never updates

**Comment by Alpvax** [comment]
There's a lot of stuff in here now that has nothing to do with items

**Comment by Trymunx** [comment]
> There's a lot of stuff in here now that has nothing to do with items

Uh oh, @Alpvax is still watching this PR 😱
I should've merged the branch first, then continued working with the changes. But I hadn't got round to fixing that disgusting surroundings getter in store.js, and I didn't want to merge before then. You're right though, this is terrible source control etiquette 😄

File: src/vuex/store.js [hunk 7]
```diff
@@ -178,33 +228,12 @@ const store = new Vuex.Store({
...
       };
-      state.player = Object.assign({}, state.player, {pos});
-    },
-    MOVE_PLAYER_RIGHT(state) {
-      const pos = {
-        x: state.player.pos.x + 1,
-        y: state.player.pos.y,
-      };
-      state.player = Object.assign({}, state.player, {pos});
+      state.player = Object.assign({}, state.player, { pos });
```
Comment (Alpvax) [inline src/vuex/store.js]: Why are you assigning a copy of the player (with a new pos) to the player var? Why not just modify the player? `Object.assign(state.player, { pos });`

Comment (Trymunx) [reply src/vuex/store.js]: I had been hit by a spate of immutability. But only for this mutation, not for others... :grimacing:

Comment (Alpvax) [reply src/vuex/store.js]: Yeah, immutability is most useful in threaded environments, and only when an object is always immutable
